File: chapter_2/exercise_2_1/exercise_2_1_love2d/mover.lua
local vector = require('./vector')

local Mover = {}
Mover.__index = Mover

function Mover.new(location)
    local t = {}
    setmetatable(t, Mover)

    t.location = location
    t.velocity = vector.new(0, 0)
    t.acceleration = vector.new(0, 0)

    t.width = 15
    t.height = 25

    t.atTop = false

    return t
end

function Mover:draw()
  -- stroke(0, 0, 0)
  love.graphics.setColor(100, 100, 100)

  love.graphics.ellipse('fill', self.location.x, self.location.y, self.width, self.height)
end

function Mover:update()
  self.velocity = self.velocity + self.acceleration
  self.location = self.location + self.velocity
  self.acceleration = self.acceleration * 0
  self.velocity = self.velocity * 0

  if self.location.y - (self.height / 2) - 10 <= 0 then
    self.atTop = true
  else
    self.atTop = false
  end
end

function Mover:applyForce(force)
  self.acceleration = self.acceleration + force
end

return Mover

File: chapter_3/example_3_3_rust/src/mover.rs
use bbggez::color::random_bright_color;
use bbggez::ggez::{graphics, nalgebra::Vector2, Context};
use bbggez::mesh::create_rect;
use bbggez::traits::NatureOfCode;

pub struct Mover {
	pub location: Vector2<f32>,
	pub mesh: graphics::Mesh,
	pub rotation: f32,
	pub velocity: Vector2<f32>,
	acceleration: Vector2<f32>,
}

impl NatureOfCode for Mover {
	fn new(x: f32, y: f32, _mass: f32, context: &mut Context) -> Mover {
		let width = 50.0;
		let height = 15.0;
		let mesh = create_rect(
			0.0 - width / 2.0,
			0.0 - height / 2.0,
			width,
			height,
			random_bright_color(),
			context,
		);

		Mover {
			location: Vector2::new(x, y),
			mesh,
			velocity: Vector2::new(0.0, 0.0),
			rotation: 0.0,
			acceleration: Vector2::new(0.0, 0.0),
		}
	}

	fn update(&mut self, delta_time: f32) {
		self.rotation = self.velocity.y.atan2(self.velocity.x);
		self.velocity += self.acceleration;
		self.location += self.velocity * delta_time;
		self.acceleration *= 0.0;
	}

	fn apply_force(&mut self, force: Vector2<f32>) {
		self.acceleration += force;
	}

	fn reverse_horizontal_velocity(&mut self) {}

	fn reverse_vertical_velocity(&mut self) {}
}
